topics/ldap.md

```diff
@@ -0,0 +1,212 @@
+
+Here’s an example of how to use `search+bind` mode for authentication.
+
+Consider the following LDAP user entry:
+
+```
+dn: cn=Ben Alex,ou=engineering,dc=valkey,dc=io
+objectClass: person
+objectClass: inetOrgPerson
+entryDN: cn=Ben Alex,ou=engineering,dc=valkey,dc=io
+cn: Ben Alex
+sn: Alex
+uid: ben
+```
+
+In this example, the username `ben` is stored in the `uid` attribute, and the user's DN is stored in the `entryDN` attribute. Since it's not possible to construct the DN using the username with a prefix and suffix as in the `bind` mode, you need to use `search+bind` mode to locate the user entry and extract its DN.
+
+Set `ldap.search_attribute` to `uid` and `ldap.search_dn_attribute` to `entryDN` to match the username and retrieve the DN.
+
+You can also use the `ldap.search_filter` option to restrict which LDAP objects are included in the search. By default, `ldap.search_filter` is set to `objectClass=*`, which matches all objects.
+
+Internally, the module performs an LDAP search with a filter like `(&(objectClass=*)(uid=ben))`, which matches the entry above. It then extracts the DN from the `entryDN` attribute and re-binds using this DN and the password provided in the `AUTH` command to complete authentication.
+
+### Comparison with Bind Mode
+
+While `bind` mode is simpler and faster, as it directly constructs the DN from the username, it is best suited for environments where the directory structure is straightforward and usernames directly correspond to DN components. 
+
+On the other hand, `search+bind` mode provides greater flexibility, allowing authentication in complex directory structures where usernames may not directly map to DN components. 
+
+However, this flexibility comes at the cost of additional LDAP requests, making it slightly slower than `bind` mode.
+
+## Setting Up Valkey Users
+
+The LDAP server is used solely for authenticating users. Authorization rules for each user must be configured in Valkey using either the `ACL SETUSER` command, through entries in `valkey.conf`, or using an [ACL file](../acl). This means that user accounts must exist in Valkey in order for LDAP authentication to succeed.
+
+For example, to allow the user `bob` to authenticate via LDAP, a corresponding user named `bob` must be present in the Valkey ACL database.
+
+To ensure that `bob` cannot log in using password-based authentication, you can create the Valkey user without specifying a password.
+
+To create a user without a password, simply omit the password when running the `ACL SETUSER` command and define the appropriate ACL rules. For example:
+
+```
+ACL SETUSER bob on +@hash
+```
+
+After creating the user `bob` in Valkey, authentication will only be possible through the LDAP module.
```

Comment:
It would be nice if there was some way to restrict a user to just a given module. Not something we have today, but something we might want to consider in the future.